apps/ejabberd/src/mongoose_api_backend.erl

```diff
@@ -0,0 +1,234 @@
+
+content_types_provided(Req, State) ->
+    CTP = [{{<<"application">>, <<"json">>, '*'}, to_json}],
+    {CTP, Req, State}.
+
+content_types_accepted(Req, State) ->
+    CTA = [{{<<"application">>, <<"json">>, '*'}, from_json}],
+    {CTA, Req, State}.
+
+rest_terminate(_Req, _State) ->
+    ok.
+
+allow_missing_post(Req, State)->
+    {true, Req, State}.
+
+%%--------------------------------------------------------------------
+%% internal callbacks
+%%--------------------------------------------------------------------
+
+to_json(Req, State) ->
+    handle_request(mongoose_api_json, Req, State).
+
+%%--------------------------------------------------------------------
+from_json(Req, State) ->
+    handle_body(mongoose_api_json, Req, State).
+
+%%--------------------------------------------------------------------
+%% internal funs
+%%--------------------------------------------------------------------
+
+handle_request(Serializer, Req, #backend_state{bindings=Bindings, command_category = Name}=State) ->
+    %% should we introduce filtering not only by category?
+    Commands = ?COMMANDS_ENGINE:list(admin, Name),
+    [Command] = [C || C <- Commands, ?COMMANDS_ENGINE:action(C) =:= read],
+    Result = execute_command(extract_bindings(Bindings), Command),
+    handle_result(Result, Serializer, Req, State).
+
+handle_body(Deserializer, Req, State) ->
+    {Method, Req1} = cowboy_req:method(Req),
+    {ok, Body, Req2} = cowboy_req:body(Req1),
+    case Deserializer:deserialize(Body) of
+        {ok, Data} ->
+            handle_body(Method, Data, Req2, State);
+        {error, _Reason} ->
+            error_response(bad_request, Req2, State)
+    end.
+
+handle_body(Method, Data, Req, #backend_state{command_category = Category}=State) ->
+    Action = method_to_action(Method),
+    Commands = ?COMMANDS_ENGINE:list(admin, Category),
+    [Command] = [C || C <- Commands, ?COMMANDS_ENGINE:action(C) =:= Action],
+    case check_and_extract_args(?COMMANDS_ENGINE:args(Command), Data) of
+        {ok, Args} ->
+            Result = execute_command(Args, Command),
+            handle_result(Result, Req, State);
+        {error, Type} ->
+            error_response(Type, Req, State)
+    end.
+
+handle_result({ok, Result}, Serializer, Req, State) ->
+    serialize(Result, Serializer, Req, State);
+handle_result(Other, _Serializer, Req, State) ->
+    handle_result(Other, Req, State).
+
+handle_result({ok, _Res}, Req, State) ->
+    %% TODO When POST add resource created to headers
+    %% TODO return appropriate status code
+    {ok, Req2} = cowboy_req:reply(201, [{<<"location">>, <<"http://localhost:5288/api/user">>}], Req),
+    {halt, Req2, State};
+handle_result({error, Error}, Req, State) ->
+    error_response(Error, Req, State);
+handle_result(no_call, Req, State) ->
+    error_response(not_implemented, Req, State).
+
+serialize(Data, Serializer, Req, State) ->
+    {Serializer:serialize(Data), Req, State}.
+
+register_handler(Base, Command) ->
```

Comment:
Does this method really register a handler, or just returns a path? If the latter then this is not a proper name.